packages/app/src/app/overmind/namespaces/editor/actions.ts

```diff
@@ -167,41 +166,41 @@ export const codeChanged: Action<{
   });
 };
 
-export const saveClicked: AsyncAction = async ({ state, effects, actions }) => {
-  await actions.editor.internal.ensureSandboxIsOwned();
-
-  const sandbox = state.editor.currentSandbox;
-  const currentlyChangedModuleShortids = state.editor.changedModuleShortids.slice();
+export const saveClicked: AsyncAction = withOwnedSandbox(
+  async ({ state, effects, actions }) => {
+    const sandbox = state.editor.currentSandbox;
+    const currentlyChangedModuleShortids = state.editor.changedModuleShortids.slice();
 
-  try {
-    const changedModules = sandbox.modules.filter(module =>
-      state.editor.changedModuleShortids.includes(module.shortid)
-    );
+    try {
+      const changedModules = sandbox.modules.filter(module =>
+        state.editor.changedModuleShortids.includes(module.shortid)
+      );
 
-    state.editor.changedModuleShortids = [];
+      state.editor.changedModuleShortids = [];
 
-    await effects.api.saveModules(sandbox.id, changedModules);
-    effects.moduleRecover.clearSandbox(sandbox.id);
+      await effects.api.saveModules(sandbox.id, changedModules);
+      effects.moduleRecover.clearSandbox(sandbox.id);
 
-    if (
-      state.editor.currentSandbox.originalGit &&
-      state.workspace.openedWorkspaceItem === 'github'
-    ) {
-      actions.git.fetchGitChanges();
-    }
-  } catch (error) {
-    // Put back any unsaved modules taking into account that you
-    // might have changed some modules waiting for saving
-    currentlyChangedModuleShortids.forEach(moduleShortid => {
-      if (!state.editor.changedModuleShortids.includes(moduleShortid)) {
-        state.editor.changedModuleShortids.push(moduleShortid);
+      if (
+        state.editor.currentSandbox.originalGit &&
+        state.workspace.openedWorkspaceItem === 'github'
+      ) {
+        actions.git.fetchGitChanges();
       }
-    });
-    effects.notificationToast.error(
-      'Sorry, was not able to save, please try again'
-    );
+    } catch (error) {
+      // Put back any unsaved modules taking into account that you
+      // might have changed some modules waiting for saving
+      currentlyChangedModuleShortids.forEach(moduleShortid => {
+        if (!state.editor.changedModuleShortids.includes(moduleShortid)) {
+          state.editor.changedModuleShortids.push(moduleShortid);
```

Comment:
Don't mutate the `state.editor.changedModuleShortids` array, give it a new value instead.

```ts
state.editor.changedModuleShortids =  [...state.editor.changedModuleShortids, moduleShortid];
```

Reply:
With Overmind we mutate as it tells the components exactly what we did, improving performance :)

Reply:
Not that it practically matters here really, as a **push**, **splice** etc. also causes a "change to the array". But just wanted to mention that we do not depend on reference checking (immutability) in either Cerebral (mobx) or Overmind :)

Reply:
It's more a best practice to not mutate variables, but instead change the total value instead.
This _can_ cause bugs when we do mutate.

Reply:
Mutations in general is bad, but the state of Overmind is like an API. Everything is controlled and tracked :) This is why Redux has performance issues in complex UIs. It is unable to pinpoint exact changes, as any change causes change to all parent objects, meaning all components looking at state at any level of the same hierarchy will need to reconcile.

I really like the simplicity of immutability and reference checking for React, but it just does not work in complex UIs as it causes performance issues without a lot of manual props checking, normalizing state and stuff.

Mobx / Mobx-State-Tree also uses controlled mutability. I think fundamentally Redux (immutability) is doing the right thing, but the API is hell to work with as JavaScript has no immutable native API and JavaScript/React is not fast enough to manage a full reconcile on any change. You have to do a lot of React.memo etc. to manage performance.

Wow, that got me going! :) Anyways, this specific change does not matter for performance, except that you will not see in the devtools what you actually pushed to the array as that information is lost.

Reply:
I understand the reasoning behind it 🙂